### PR TITLE
test(lib): ensure list mapper does not fail when passed IResolvables

### LINKED ISSUE #835: Property Mapper should Respect Tokens
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

- all

### Affected Resource(s)

- Lazy / Token

### Debug Output

```
node_modules/cdktf/lib/runtime.js:16
            return x.map(elementMapper);
                     ^

    TypeError: x.map is not a function
```

### Expected Behavior

`IResolvable` should be skipped here as well https://github.com/hashicorp/terraform-cdk/blob/d72fc072d3e6cf0a27ddef135942bb32f5ee8eb5/packages/cdktf/lib/runtime.ts#L50

### Actual Behavior

It's trying to map `IResolvable` values, which fails.

### LINKED ISSUE #910: bug(lib): Fix Token Resolve
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

- all

### Affected Resource(s)

- Tokens

### Description

I took some time to compare the synth process of AWS CDK and our implementation. At a first glance, it works similar

#### CfnStack

The AWS CDK stack calls `_toCloudformation` on [all descendant](https://github.com/aws/aws-cdk/blob/ab228a648fa08e8189c010d475426f1ff003ddf7/packages/@aws-cdk/core/lib/stack.ts#L982-L983) descendants and resolves each element individually.

```ts
    const elements = cfnElements(this);
    const fragments = elements.map(e => this.resolve(e._toCloudFormation()));
```

#### TerraformStack

That's more or less what we're doing as well:

https://github.com/hashicorp/terraform-cdk/blob/720d6a782f553d7f1fefb3b22fb84ed43e0cd5a9/packages/cdktf/lib/terraform-stack.ts#L164-L167

#### CfnResource

However, the interesting part happens in [_toCloudFormation](https://github.com/aws/aws-cdk/blob/ebac8bc08885d6862f75b1133752b639dcf54b1c/packages/@aws-cdk/core/lib/cfn-resource.ts#L313) of the `CfnResource`. There we come across the [PostResolveToken](), which essentially postpones the rendering of the properties to a [post processor](https://github.com/aws/aws-cdk/blob/ebac8bc08885d6862f75b1133752b639dcf54b1c/packages/@aws-cdk/core/lib/cfn-resource.ts#L325). 

```ts
          [this.logicalId]: new PostResolveToken({
            Type: this.cfnResourceType,
            Properties: ignoreEmpty(this.cfnProperties),
            DependsOn: ignoreEmpty(renderDependsOn(this.dependsOn)),
            CreationPolicy: capitalizePropertyNames(this, renderCreationPolicy(this.cfnOptions.creationPolicy)),
            UpdatePolicy: capitalizePropertyNames(this, this.cfnOptions.updatePolicy),
            UpdateReplacePolicy: capitalizePropertyNames(this, this.cfnOptions.updateReplacePolicy),
            DeletionPolicy: capitalizePropertyNames(this, this.cfnOptions.deletionPolicy),
            Version: this.cfnOptions.version,
            Description: this.cfnOptions.description,
            Metadata: ignoreEmpty(this.cfnOptions.metadata),
            Condition: this.cfnOptions.condition && this.cfnOptions.condition.logicalId,
          }, resourceDef => {
            const renderedProps = this.renderProperties(resourceDef.Properties || {});
            if (renderedProps) {
              const hasDefined = Object.values(renderedProps).find(v => v !== undefined);
              resourceDef.Properties = hasDefined !== undefined ? renderedProps : undefined;
            }
            const resolvedRawOverrides = Tokenization.resolve(this.rawOverrides, {
              scope: this,
              resolver: CLOUDFORMATION_TOKEN_RESOLVER,
              // we need to preserve the empty elements here,
              // as that's how removing overrides are represented as
              removeEmpty: false,
            });
            return deepMerge(resourceDef, resolvedRawOverrides);
          }),
```
#### TerraformResource

In contrast to this, we're rendering the properties right away 

https://github.com/hashicorp/terraform-cdk/blob/720d6a782f553d7f1fefb3b22fb84ed43e0cd5a9/packages/cdktf/lib/terraform-resource.ts#L112-L128

### Conclusion 

I'm rather confident, that this difference is causing some of the subtile issues we're seeing, e.g. https://github.com/hashicorp/terraform-cdk/pull/836

### LINKED ISSUE #1788: [Research] Token Resolution Issue
Placeholder to do some research into #910 and the original source issue, #835. 

Verify (a) that it's still an active issue, and (b) whether what Sebastian identified in #910 does seem to be the cause.

We'll do the research in 0.11 so that any implementation work can happen in 0.12.

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Closes #1788
Closes #910 
Closes #835

### CODE REVIEW

**Review by ansgarm — APPROVED** [review]
[.](https://www.youtube.com/watch?v=UBX8MWYel3s)

**Comment by jsteinich** [comment]
I believe #835 has been fixed, but I'm not yet convinced that #910 still doesn't cause issues.

I haven't quite thought of a case where this would actually happen, but if `deepMerge` ends up mixing tokens and non-tokens within the same property the results could be wrong.
Another spot worth testing is `Lazy`. We don't really use it, but it does exist.

**Comment by DanielMSchmidt** [comment]
>  if deepMerge ends up mixing tokens and non-tokens within the same property the results could be wrong.

on the same property as in different types of tokens in a list or as function invocations?

**Comment by jsteinich** [comment]
> > if deepMerge ends up mixing tokens and non-tokens within the same property the results could be wrong.
> 
> on the same property as in different types of tokens in a list or as function invocations?

A couple possible cases:
* Trying to merge a token and non-token. One direct just overwrites, the other might append to the token object.
* Adding token and non-token to list. Resolution will fail later.

@ansgarm there was a slack thread recently about JSON.stringify; would that line up with this.

File: packages/cdktf/lib/util.ts [hunk 1]
```diff
@@ -5,6 +5,14 @@ import { Tokenization } from "./tokens/token";
  */
 
 export function deepMerge(target: any, ...sources: any[]) {
+  if (Tokenization.isResolvable(target) && sources.length > 0) {
+    throw new Error(
+      `Invalid usage. Target (${JSON.stringify(
```
Comment (jsteinich) [inline packages/cdktf/lib/util.ts:10]: I wonder if just using `target.toString` would be more readable.

**Pull request merged**